drivers/net/nsos_sockets.c

```diff
@@ -0,0 +1,525 @@
+{
+	struct nsos_mid_sockaddr_storage addr_storage_mid;
+	struct nsos_mid_sockaddr *addr_mid = (struct nsos_mid_sockaddr *)&addr_storage_mid;
+	size_t addrlen_mid;
+	int ret;
+
+	ret = sockaddr_to_nsos_mid(addr, &addrlen, &addr_mid, &addrlen_mid);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	ret = nsos_adapt_bind(OBJ_TO_FD(obj), addr_mid, addrlen_mid);
+
+return_ret:
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static int nsos_connect(void *obj, const struct sockaddr *addr, socklen_t addrlen)
+{
+	struct nsos_mid_sockaddr_storage addr_storage_mid;
+	struct nsos_mid_sockaddr *addr_mid = (struct nsos_mid_sockaddr *)&addr_storage_mid;
+	size_t addrlen_mid;
+	int ret;
+
+	ret = sockaddr_to_nsos_mid(addr, &addrlen, &addr_mid, &addrlen_mid);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	ret = nsos_adapt_connect(OBJ_TO_FD(obj), addr_mid, addrlen_mid);
+
+return_ret:
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static int nsos_listen(void *obj, int backlog)
+{
+	int ret;
+
+	ret = nsos_adapt_listen(OBJ_TO_FD(obj), backlog);
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static int nsos_accept(void *obj, struct sockaddr *addr, socklen_t *addrlen)
+{
+	struct nsos_mid_sockaddr_storage addr_storage_mid;
+	struct nsos_mid_sockaddr *addr_mid = (struct nsos_mid_sockaddr *)&addr_storage_mid;
+	size_t addrlen_mid = sizeof(addr_storage_mid);
+	int ret;
+
+	ret = nsos_adapt_accept(OBJ_TO_FD(obj), addr_mid, &addrlen_mid);
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	adapt_fd = ret;
+
+	ret = sockaddr_from_nsos_mid(addr, addrlen, addr_mid, addrlen_mid);
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		goto close_adapt_fd;
+	}
+
+	zephyr_fd = z_reserve_fd();
+	if (zephyr_fd < 0) {
+		goto close_adapt_fd;
+	}
+
+	z_finalize_fd(zephyr_fd, FD_TO_OBJ(adapt_fd), &nsos_socket_fd_op_vtable.fd_vtable);
+
+	return zephyr_fd;
+
+close_adapt_fd:
+	nsi_host_close(adapt_fd);
+
+	return -1;
+}
+
+static ssize_t nsos_sendto(void *obj, const void *buf, size_t len, int flags,
+			   const struct sockaddr *addr, socklen_t addrlen)
+{
+	struct nsos_mid_sockaddr_storage addr_storage_mid;
+	struct nsos_mid_sockaddr *addr_mid = (struct nsos_mid_sockaddr *)&addr_storage_mid;
+	size_t addrlen_mid = sizeof(addr_storage_mid);
+	int flags_mid;
+	int ret;
+
+	ret = socket_flags_to_nsos_mid(flags);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	flags_mid = ret;
+
+	ret = sockaddr_to_nsos_mid(addr, &addrlen, &addr_mid, &addrlen_mid);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	ret = nsos_adapt_sendto(OBJ_TO_FD(obj), buf, len, flags_mid,
+				addr_mid, addrlen_mid);
+
+return_ret:
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static ssize_t nsos_sendmsg(void *obj, const struct msghdr *msg, int flags)
```

Comment:
We currently also support `recvmsg` so please add `nsos_recvmsg` and return `ENOTSUP` for it too

Reply:
Added `nsos_recvmsg`, as suggested.